### ISSUE TITLE
Publish Events on Step & Task Completions

### ISSUE THREAD

**Issue by mustafaakin** [tektoncd/pipeline#742]
# Expected Behavior

Have more knowledge of what happened in the PipelineRun or TaskRuns

# Actual Behavior

There are only Succeeded or Failed events for whole Pipeline or Conditions for TaskRun/PipelineRun.

# Additional Info

As the containers in the pod run in parallel and they are all starting at the same time and ending in arbitrary times, it can be hard to debug what went. Publishing events on when tasks finished can be useful to users.

**Comment by afrittoli** [comment]
We do publish some events today, but only in case of errors and afaik only for Tasks and Pipelines. There may be some standard events published for containers. 

We do have start and stop times of everything in the status, so with that it is possible to build a timeline of execution - perhaps displaying the timeline is something that it would be nice to integrate into the CLI and/or the dashboard. Would this be sufficient for your use case?

**Comment by jeffmaury** [comment]
I noticed no events are generated when a tekton resource is created/deleted. Is this planned ? We need this for the VSCode and IntelliJ stuff we are working on in order to refresh the tree

**Comment by bobcatfish** [comment]
> I noticed no events are generated when a tekton resource is created/deleted. Is this planned ?

@jeffmaury do you mean kubernetes events or CloudEvents? Either way I think we could add this but I don't think we have a plan atm.

**Comment by AloisReitbauer** [comment]
The Keptn project wants to work on a joint definition of these events. See issue in Keptn: https://github.com/tektoncd/pipeline/issues/742

**Comment by sladyn98** [comment]
@bobcatfish I would like to get started on this issue as it is marked as a good-first issue, any pointers to get started.

**Comment by afrittoli** [comment]
FYI - there are some changes happening being considered about when a pipeline is considered as "complete": #1749 and #2094 .

@mustafaakin is this issue about emitting kubernetes events?
I'm going to work on having cloud events sent (optionally) upon pipeline/taskrun life-cycle events, but I think that is a separate issue.

**Comment by sladyn98** [comment]
@afrittoli Thanks, so should I pick up another issue :)
